Modifications to the  wrfarw souce code (in dyn_em/solve_em.F) to avoid model crash in updatevars step

GangZhao-NOAA opened the pull request:
1. modified wrfarw souce code (in dyn_em/solve_em.F)
  add a stop before model forecast (following GSL Guoqing Ge's suggestion)
  to avoid possible model crash (for 3DRTMA only)
2. when compiling wrfarwe model, first compile with original solve_em.orig.F,
  then compile with modifiled sovle_em.nofcst.F. Then two wrf.exe would be
  generated.
3. if using the modified wrfarw model, then do not use "quilt" in model run,
  so use the "No Quilt" namelist file (hrrr_conus.noQuilt.nl) in updatevars.ksh

GangZhao-NOAA committed:
1.modified wrfarw souce code (in dyn_em/solve_em.F)
  add a stop before model forecast (following GSL Guoqing Ge's suggestion)
  to avoid possible model crash (for 3DRTMA only)
2.when compiling wrfarwe model, first compile with original solve_em.orig.F,
  then compile with modifiled sovle_em.nofcst.F. Then two wrf.exe would be
  generated.
3.if using the modified wrfarw model, then do not use "quilt" in model run,
  so use the "No Quilt" namelist file (hrrr_conus.noQuilt.nl) in updatevars.ksh

MatthewMorris-NOAA commented on sorc/install_rtma3d_all.sh:
```diff
@@ -28,7 +28,9 @@ fi
 if [ $INSTALL_rtma3d_wrfarw = yes ] ; then
 
-cp -p ${BASE}/rtma3d_wrfarw.fd/WRFV3.9/main/wrf.exe ../exec/rtma3d_wrfarw_fcst
+# cp -p ${BASE}/rtma3d_wrfarw.fd/WRFV3.9/main/wrf.exe ../exec/rtma3d_wrfarw_fcst
```
Can these be renamed for simplicity? E.g., keep the original file name "rtma3d_wrfarw_fcst" and the executable with no model forecast would then be named as "rtma3d_wrfarw_no_fcst".

GangZhao-NOAA replied:
@MatthewMorris-NOAA 
I followed your suggestion and modified the filename of the wrf exe in scripts.
Thank you!

MatthewMorris-NOAA replied:
Great, thanks @GangZhao-NOAA!

MatthewMorris-NOAA commented on sorc/build_rtma3d_wrfarw.sh:
```diff
@@ -22,7 +22,26 @@ cd ${BASE}/rtma3d_wrfarw.fd/WRFV3.9
+
+# 1. compiling with original dyn_em/solve_em.orig.F (as a backup)
+cp -p configure.wrf.useme configure.wrf
+cp -p dyn_em/ dyn_em/solve_em.orig.F dyn_em/solve_em.F
```
This seems to be extraneous.
cp -p **dyn_em/** dyn_em/solve_em.orig.F dyn_em/solve_em.F

GangZhao-NOAA replied:
@MatthewMorris-NOAA This is a bug. Thank you for pointing it out!
It was fixed in the latest commit. See the updated PR #79.

MatthewMorris-NOAA commented on scripts/rtma3d/exrtma3d_updatevars.ksh:
```diff
@@ -94,7 +94,15 @@ cd ${DATAHOME}
 export WRF_NAMELIST=${DATAHOME}/namelist.input
-${CP} ${PARMwrf}/hrrr_conus.nl ${WRF_NAMELIST} 
+
+export L_WRFARW_NOFCST=${L_WRFARW_NOFCST:-"Yes"}         # using the modified model than does not do the integral (do not use quilt in namelist)
```
Will this be defined upstream, e.g., in the workflow XML files?

guoqing-noaa commented:
@GangZhao-NOAA Do we need to run wrf integrations in 3DRTMA? If not, we don't need to build two executables.

MatthewMorris-NOAA reviewed (commented):
@GangZhao-NOAA The changes generally look OK, but I left a few comments. It would also be preferable to include (and test) the same changes for the Alaska domain before approving and merging this PR. Thanks!

GangZhao-NOAA committed:
Using the modified WRF code for 3DRTMA run on Alaska domain,
 similarly as for run on CONUS domain.

GangZhao-NOAA committed:
fixed a bug in build_rtma3d_wrfarw.sh
simplified the wrf exe filenames used in updatevars scripts and install script.

GangZhao-NOAA commented:
Hi @guoqing-noaa 
> @GangZhao-NOAA Do we need to run wrf integrations in 3DRTMA? If not, we don't need to build two executables.

I agree with you. Just considering that hrrr-3drtma is still in the stage of development and testing, in case that the udpatevars step crashes even with the modified wrf code, we can use the original wrf code to test, so two wrf exe files are compiled here.

This is easy to get user confused. So definitely when delivering the released version to NCO, only one wrf exe file would be compiled, installed and used.

Thank you!

guoqing-noaa commented:
> Hi @guoqing-noaa
> 
> > @GangZhao-NOAA Do we need to run wrf integrations in 3DRTMA? If not, we don't need to build two executables.
> 
> I agree with you. Just considering that hrrr-3drtma is still in the stage of development and testing, in case that the udpatevars step crashes even with the modified wrf code, we can use the original wrf code to test, so two wrf exe files are compiled here.
> 
> This is easy to get user confused. So definitely when delivering the released version to NCO, only one wrf exe file would be compiled, installed and used.
> 
> Thank you!

Thanks for sharing the considerations. But since we don't do any model integration (we stopped before the starting of any integrations), it makes sense NOT to compile two wrf.exe. If there is a potential crash, we need to debug based on the no-integration version.

AnnetteGibbs-NOAA reviewed (approved):
Looks good!  I approve!

MatthewMorris-NOAA reviewed (approved):
@GangZhao-NOAA The changes look good. I will apply these changes to the parallels on Monday (or once the PR is merged).

GangZhao-NOAA commented:
> > Hi @guoqing-noaa
> > > @GangZhao-NOAA Do we need to run wrf integrations in 3DRTMA? If not, we don't need to build two executables.
> > 
> > 
> > I agree with you. Just considering that hrrr-3drtma is still in the stage of development and testing, in case that the udpatevars step crashes even with the modified wrf code, we can use the original wrf code to test, so two wrf exe files are compiled here.
> > This is easy to get user confused. So definitely when delivering the released version to NCO, only one wrf exe file would be compiled, installed and used.
> > Thank you!
> 
> Thanks for sharing the considerations. But since we don't do any model integration (we stopped before the starting of any integrations), it makes sense NOT to compile two wrf.exe. If there is a potential crash, we need to debug based on the no-integration version.
Hi @guoqing-noaa,
As you know, compiling wrf it takes quite a while (about one hour or longer). So we would like to compile two wrf executables once when first time installing the package. Then if in any case we need to use the original code to verify the results, we do not need to compile the code. Definitely in the final release, there will be no need to compile the original code. 
Thank you!
-Gang

GangZhao-NOAA commented:
> @GangZhao-NOAA The changes look good. I will apply these changes to the parallels on Monday (or once the PR is merged).

Hi @MatthewMorris-NOAA ,

If you are going to use this updated package, remember to copy the updated wrf namelist files under "**_/lfs/h2/emc/da/noscrub/gang.zhao/packages/rtma3d.v1.0.0/parm/wrf2_**" to your parm/wrf/ directory.

Thank you!

-Gang
